### ISSUE TITLE
CodeGen: Task cancellation is not propagated from grains to callers

### ISSUE THREAD

**Issue by ReubenBond** [dotnet/orleans#847]
If returns a `Task` which completes with a status of `TaskStatus.Canceled`, that cancelation is ignored and the grain method call appears to complete successfully.

For example, a call to a grain with a method like this will not throw:
```c#
public Task Cancelled()
{
    var tcs = new TaskCompletionSource<int>();
    tcs.TrySetCanceled();
    return tcs.Task;
}
```

The root cause is that code generation for grain methods returning `Task` (vs `Task<T>`) does not cover the cancellation case. Example of current generated code:
```c#
return ((IExceptionGrain)grain).Cancelled().ContinueWith(t => {
  if (t.Status == System.Threading.Tasks.TaskStatus.Faulted)
    throw t.Exception;
  return (object)null;
});
```

Repro in #846 
Fixed in #528

**Comment by gabikliot** [comment]
That is a great catch and repro Reuben! Especially since it happens only in `Task` and not `Task<T>`. Can you point to the exact fix line in #528? So I can take a look at how it is fixed there?

**Comment by ReubenBond** [comment]
The precise fix is in `TaskExtensions.cs`. Instead of `.ContinueWith(...)`, #528 uses a `Box` extension method which serves the same purpose, but is implemented as follows:
```c#
public static Task<object> Box(this Task task)
{
    return task.ContinueWith(
        antecedent => default(object),
        TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.OnlyOnRanToCompletion);
}

public static Task<object> Box<T>(this Task<T> task)
{
    return task.ContinueWith(
        antecedent => (object)antecedent.Result,
        TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.OnlyOnRanToCompletion);
}
```

**Comment by gabikliot** [comment]
So `TaskContinuationOptions.OnlyOnRanToCompletion` does NOT result in cancelled exception if the task actually fails and the CW does not run? The original exception is re-thrown?

**Comment by ReubenBond** [comment]
`OnlyOnRanToCompletion` avoids altering an exception stack trace & has the added benefit that cancellations are also propagated directly through (monadically). A canceled task will throw an exception during `await`/`.Result`/`.Wait()`, but you can check the cancellation status of a task without an exception being raised.

**Comment by jdom** [comment]
As explained in #853, we should not use TaskContinuationOptions.OnlyOnRanToCompletion, since it will actually cancel the continuation task if the precursor task is faulted or canceled.
I believe what you are looking at to preserve the stack is
task.ContinueWith(antecedent => (object)antecedent.GetAwaiter().GetResult())

You could also potentially use the ExecuteSynchronously (it does seem appropriate, but I don't have all the context here to know if that is the case), but definitely not the OnlyRanToCompletion

**Comment by ReubenBond** [comment]
@jdom, I agree and should have seen that. What it the difference between `.Result` and `.GetAwaiter().GetResult()`?
I'd consider passing exceptions and cancellations through directly using a `TaskCompletionSource<object>` because of stack preservation.

**Comment by jdom** [comment]
The main difference is that in failure scenarios `.Result` always throws an `AggregateException` that wraps the original exception, where as `.GetAwaiter().GetResult()` throws the original exception, but also makes sure to preserve the original stack trace too (plus add a line in the stack trace about were it threw of course). This stack trace looks effectively the same as if we used `await antecedent`

**Comment by jdom** [comment]
BTW, some sources:
https://github.com/aspnet/Security/issues/59
http://stackoverflow.com/questions/17284517/is-task-result-the-same-as-getawaiter-getresult

And also a simple repro I wrote to verify this:

        [TestMethod]
        public async Task TestMethod2()
        {
            Task<int> task = Task.Run<int>(() =>
            {
                Console.WriteLine("Returning 42");
                throw new InvalidOperationException();
                return 42;
            });

            var continuation = task.ContinueWith(prevTask =>
            {
                var prevValue = prevTask.GetAwaiter().GetResult();
                return prevValue;
            });

            try
            {
                var value = await continuation;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.ToString());
            }
        }

And the output from this test:

		Returning 42
		System.InvalidOperationException: Operation is not valid due to the current state of the object.
		   at TasksContinuations.UnitTest1.<TestMethod2>b__4() in d:\temp\TasksContinuations\TasksContinuations\UnitTest1.cs:line 34
		   at System.Threading.Tasks.Task`1.InnerInvoke()
		   at System.Threading.Tasks.Task.Execute()
		--- End of stack trace from previous location where exception was thrown ---
		   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
		   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
		   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
		   at TasksContinuations.UnitTest1.<TestMethod2>b__5(Task`1 prevTask) in d:\temp\TasksContinuations\TasksContinuations\UnitTest1.cs:line 40
		   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
		   at System.Threading.Tasks.Task.Execute()
		--- End of stack trace from previous location where exception was thrown ---
		   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
		   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
		   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
		   at TasksContinuations.UnitTest1.<TestMethod2>d__8.MoveNext() in d:\temp\TasksContinuations\TasksContinuations\UnitTest1.cs:line 46

**Comment by gabikliot** [comment]
:+1: Thanks Julian for clarifying that!

**Comment by ReubenBond** [comment]
:+1: Thanks for the explanation, @jdom!

**Comment by ReubenBond** [comment]
Resolved by #528
